EventFilter/RPCRawToDigi/plugins/RPCDigiMerger.cc

```diff
@@ -0,0 +1,102 @@
+#include "EventFilter/RPCRawToDigi/plugins/RPCDigiMerger.h"
+
+#include <memory>
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/CRC16.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "DataFormats/RPCDigi/interface/RPCDigiCollection.h"
+
+using namespace edm;
+using namespace std;
+
+RPCDigiMerger::RPCDigiMerger(edm::ParameterSet const & config)
+{
+    produces<RPCDigiCollection>();
+    // if (fill_counters_) {
+    //     produces<RPCAMCLinkCounters>();
+    // }
+    TwinMux_token_ = consumes<RPCDigiCollection>(config.getParameter<edm::InputTag>("inputTagTwinMuxDigis"));
+    OMTF_token_ = consumes<RPCDigiCollection>(config.getParameter<edm::InputTag>("inputTagOMTFDigis"));
+    CPPF_token_ = consumes<RPCDigiCollection>(config.getParameter<edm::InputTag>("inputTagCPPFDigis"));
+}
+
+RPCDigiMerger::~RPCDigiMerger()
+{}
+
+void RPCDigiMerger::fillDescriptions(edm::ConfigurationDescriptions & descs)
+{
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("inputTagTwinMuxDigis", edm::InputTag("RPCTwinMuxRawToDigi", ""));
+    desc.add<edm::InputTag>("inputTagOMTFDigis", edm::InputTag("omtfStage2Digis", ""));
+    desc.add<edm::InputTag>("inputTagCPPFDigis", edm::InputTag("rpcCPPFRawToDigi", ""));
+
+    descs.add("RPCDigiMerger", desc);
+}
+
+void RPCDigiMerger::beginRun(edm::Run const & run, edm::EventSetup const & setup)
+{}
+
+void RPCDigiMerger::produce(edm::Event & event, edm::EventSetup const & setup)
+{
+
+    // Get the digis 
+    // TwinMux
+    Handle<RPCDigiCollection> TwinMux_digis; 
+    event.getByToken(TwinMux_token_,TwinMux_digis);
+    // OMTF
+    Handle<RPCDigiCollection> OMTF_digis; 
+    event.getByToken(OMTF_token_,OMTF_digis);
+    // CPFF
+    Handle<RPCDigiCollection> CPPF_digis; 
+    event.getByToken(CPPF_token_,CPPF_digis);
+
+    // new RPCDigiCollection
+    std::unique_ptr<RPCDigiCollection> rpc_digi_collection(new RPCDigiCollection());
+
+    // loop over twinmux digis
+    for ( auto rpcdgIt = TwinMux_digis->begin(); rpcdgIt != TwinMux_digis->end(); ++rpcdgIt ) {
```

Comment:
A range based for loop may better apply here:
```
    for (auto & rpcdgIt : TwinMux_digis ) {
        // The layerId
        const RPCDetId& rpcId = rpcdgIt.first;
        // Get the iterators over the digis associated with this LayerId
        const RPCDigiCollection::Range& range = rpcdgIt.second;

        rpc_digi_collection->put(range, rpcId);
    }
```

Reply:
Hi @perrotta,

I can't range loop over a pointer.

```
/afs/cern.ch/work/f/ftorresd/private/RPC/RPCTM/rpcDQM/v07_GitHub_PR/CMSSW_10_3_0_pre4/src/EventFilter/RPCRawToDigi/plugins/RPCDigiMerger.cc:64:27: error: 'begin' was not declared in this scope
     for (auto & rpcdgIt : TwinMux_digis ) {
```

If I try to get TwinMux_digis by deference (*TwinMux_digis) I get:

```
error: cannot bind non-const lvalue reference of type 'std::pair<RPCDetId, std::pair<__gnu_cxx::__normal_iterator<const RPCDigi*, std::vector<RPCDigi, std::allocator<RPCDigi> > >, __gnu_cxx::__normal_iterator<const RPCDigi*, std::vector<RPCDigi, std::allocator<RPCDigi> > > > >&' to an rvalue of type 'DigiContainerIterator<RPCDetId, RPCDigi>::value_type {aka std::pair<RPCDetId, std::pair<__gnu_cxx::__normal_iterator<const RPCDigi*, std::vector<RPCDigi, std::allocator<RPCDigi> > >, __gnu_cxx::__normal_iterator<const RPCDigi*, std::vector<RPCDigi, std::allocator<RPCDigi> > > > >}'
     for (auto & rpcdgIt : (*TwinMux_digis) ) {
```

Any suggestion?

Reply:
In the second case, does const auto work? (RPCDigiCollection is not a simple stl type, but it does define begin() and end() so should be ok)

Reply:
It works. Thanks!